### ISSUE TITLE
Incorrect missing property error message for (A | B) & C.

### ISSUE THREAD

**Issue by kujon** [Microsoft/TypeScript#17529]
**TypeScript Version:**  2.4.0

**Code**
```ts
type A =
    {
        foo: number;
    } |
    {
        bar: string;
    };

type B = A & {
    baz: boolean;
};

const val: B = { foo: 42 };
```

**Expected behavior:**
Compiler correctly reports that `baz` property is missing.

**Actual behavior:**
Compiler incorrectly reports the following:
`Property 'bar' is missing in type '{ foo: number; }'`

**Comment by kujon** [comment]
It can get even more confusing that this:

```ts
type A =
    {
        foo: number;
    } |
    {
        bar: string;
    };

type B = A & {
    baz: boolean;
};

const val: B = { foo: 42, baz: 42 };
```

The example above still gives: `Property 'bar' is missing in type '{ foo: number; baz: 42; }'.`, whereas the correct error should mention type of number being incompatible with type of boolean.

**Comment by DanielRosenwasser** [comment]
@sandersn has worked on something mildly similar with discriminated unions, I think it'd be great if we could improve error messages here as well though.

**Comment by njgraf512** [comment]
Another seemingly related example that I'm currently struggling with:

```ts
type Foo = {
  foo: string,
}
type FooPlus = {
  fooPlus: Foo
}

type Bar = {
  bar: string,
}
type BarPlus = {
  barPlus: Bar
}

type Baz = {
  baz: string,
}
type BazPlus = {
  bazPlus: Baz,
}

type FooPlusBarPlusUnion = FooPlus | BarPlus;

type MyType = {
  myType: string
} & FooPlusBarPlusUnion & BazPlus

const myType: MyType = {
  myType: 'myType',
  fooPlus: {
    foo: 'foo',
  },
  bazPlus: {
    baz: 'baz',
  },
}
```

Returning error: `Property 'barPlus' is missing in type '{ myType: string; fooPlus: { foo: string; }; bazPlus: { baz: string; }; }'.`

I would expect this not to error.  Should I need to use a discriminated union here? If so, could someone explain why? Adding a discriminant key to `FooPlus` and `BarPlus` does not seem to help anyway, unfortunately.

**Comment by jack-williams** [comment]
This is the error message for the original example:
```
Type '{ foo: number; }' is not assignable to type 'B'.
  Type '{ foo: number; }' is not assignable to type '{ foo: number; } & { baz: boolean; }'.
    Property 'baz' is missing in type '{ foo: number; }' but required in type '{ baz: boolean; }'. [2322]
```
Seems like this is fixed?
